lib/core/security/profileRepository.js

```diff
@@ -361,10 +365,45 @@ class ProfileRepository extends Repository {
       }
     };
 
-    const hits = await this.module.user.search({ query }, {from: 0, size: 1});
+    if (onAssignedUsers === 'remove') {
+      const batch = [];
+      let treated = 0;
+      let userPage = await this.module.user.search(
+        { query },
+        { scroll: '1m', size: 1000 });
 
-    if (hits.total > 0) {
-      throw kerror.get('security', 'profile', 'in_use');
+      while (treated < userPage.total) {
+        batch.length = 0;
+
+        for (const user of userPage.hits) {
+          user.profileIds = user.profileIds.filter(e => {
+            return (e !== profile._id);
+          });
+          if (user.profileIds.length === 0) {
+            user.profileIds.push('anonymous');
+          }
+          batch.push(this.module.user.update(
```

Comment:
Each page contains at most 1000 users. This means we can send up to 1000 concurrent updates, I'm not sure that ES won't complain.

You probably should use smaller pages. Or implement a `mUpdate` method, if you feel like it.
Or perhaps try an [updateByQuery](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-update-by-query.html) query using a script to remove the profile value, something like that:

```js
{
  script: {
    source: `ctx._source.profileIds.remove('${profile._id}')`,
    lang: 'painless',
  },
  query,
}
```

Reply:
Gonna try the **updateByQuery** method, makes way more sense.

Reply:
I'll bring the page size to 100,
Liked the updateByQuery idea, will do with a mUpdate feature in the future.